tao_compiler/mlir/xla/ral/context/common_context_impl.h

```diff
@@ -59,19 +67,55 @@ struct RalGlobalConstantState : public tao::ral::Context::Resource {
   // in theory, there might be two constants with the same data value but
   // different shapes however for simplicity we just use the whole unique_name
   // as the key. This can be further optimized in case neccessary.
-  std::unordered_map<std::string, std::pair<buffer_t, buffer_shape_t>>
-      device_constants;
-  std::unordered_map<std::string, std::pair<buffer_t, buffer_shape_t>>
-      host_constants;
-
-  // fast path: using a const char * pointer to look up.
-  // Note that: any compiled-const should have a static global const string,
-  // thus it's safe to just use the pointer as key.
-  std::unordered_map<const char*, std::pair<buffer_t, buffer_shape_t>>
-      device_constants_by_cstr;
-  // fast path: using a const char * pointer to look up.
-  std::unordered_map<const char*, std::pair<buffer_t, buffer_shape_t>>
-      host_constants_by_cstr;
+  using Item = std::pair<buffer_t, buffer_shape_t>;
+  using ItemMap = std::unordered_map<std::string, Item>;
+  ItemMap device_constants;
+  ItemMap host_constants;
+
+  // fast path: using a unique const index to do look up.
+  // Note that the const index is assigned to each const at compile time.
+  // The index is unique within the compiled module level.
+  //
+  // The large index suppored to use the fast path.
+  static constexpr const int kMaxNumItemsForFastPath = 8196;
+  template <typename T>
+  using ItemArray = std::array<T, kMaxNumItemsForFastPath>;
+  using ItemLookupFastPathStorage = ItemArray<Item>;
+  using ItemLookupFastPathTable = ItemArray<std::atomic<Item*>>;
```

Comment:
Why need yet another array for pointer storage?

Reply:
This is used to avoid using explicit lock in the fast path. The item pointer is updated by one thread and read by other threads. To avoid read-after-write hazard, I use an atomic pointer here.